modules/apps/frontend-icons/frontend-icons-admin-web/src/main/java/com/liferay/frontend/icons/admin/web/internal/portlet/action/SaveCustomIconMVCActionCommand.java

```diff
@@ -0,0 +1,134 @@
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.icons.admin.web.internal.portlet.action;
+
+import com.liferay.configuration.admin.constants.ConfigurationAdminPortletKeys;
+import com.liferay.frontend.icons.admin.web.contributor.extender.IconResourcesContributor;
+import com.liferay.frontend.icons.admin.web.contributor.extender.internal.IconResourceImpl;
+import com.liferay.frontend.icons.admin.web.contributor.extender.internal.IconResourcesContributorImpl;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.module.configuration.ConfigurationProvider;
+import com.liferay.portal.kernel.portlet.bridges.mvc.BaseMVCActionCommand;
+import com.liferay.portal.kernel.portlet.bridges.mvc.MVCActionCommand;
+import com.liferay.portal.kernel.security.auth.PrincipalException;
+import com.liferay.portal.kernel.security.permission.PermissionChecker;
+import com.liferay.portal.kernel.security.permission.PermissionThreadLocal;
+import com.liferay.portal.kernel.servlet.SessionErrors;
+import com.liferay.portal.kernel.theme.ThemeDisplay;
+import com.liferay.portal.kernel.util.HashMapDictionaryBuilder;
+import com.liferay.portal.kernel.util.ParamUtil;
+import com.liferay.portal.kernel.util.WebKeys;
+
+import javax.portlet.ActionRequest;
+import javax.portlet.ActionResponse;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+import java.util.Dictionary;
+import java.util.Hashtable;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * @author Bryce Osterhaus
+ */
+@Component(
+	immediate = true,
+	property = {
+		"javax.portlet.name=" + ConfigurationAdminPortletKeys.INSTANCE_SETTINGS,
+		"mvc.command.name=/icons_admin/save_custom_icon"
+	},
+	service = MVCActionCommand.class
+)
+public class SaveCustomIconMVCActionCommand
+	extends BaseMVCActionCommand {
+
+	private ServiceRegistration<IconResourcesContributor> _serviceRegistration;
+
+	public void addCustomIcon(String id, String svgContent) {
+		String name = "custom_" + id;
+		String viewBox = "";
+
+		Pattern pattern = Pattern.compile("(?i)(viewBox.*\"(?=[\\s, >]))");
+		Matcher matcher = pattern.matcher(svgContent);
+		if (matcher.find()) {
+			viewBox = matcher.group(1);
+		}
+
+		svgContent = svgContent.replaceFirst(
+			"(?i)<svg[^>]*>", StringPool.BLANK);
+
+		svgContent = svgContent.replaceFirst(
+			"(?i)</svg[^>]*>", StringPool.BLANK);
+
+
+		svgContent = "<symbol id=" + StringPool.QUOTE +name+ StringPool.QUOTE + StringPool.SPACE + viewBox + ">" + svgContent + "</symbol>";
+
+		_customIconResourcesContributor.addIconResource(
+			new IconResourceImpl(name, svgContent)
+		);
```

Comment:
Rather than just storing every custom icon in the custom contributor, I think it makes more sense to off load this to store the actual svg files somewhere. Any pointers on how I should store the custom SVGs through the UI?